### ISSUE TITLE
[RFC] Support MiniCPM-o 4.5 simplex on-policy distillation

### ISSUE THREAD

**Issue by NancyFyong** [verl-project/verl-omni#565]
## Feature request

Support **MiniCPM-o 4.5 simplex on-policy distillation (OPD)** using the existing omni V1 trainer and verl distillation infrastructure.

**Parent RFC:** #345 — Real-Time Full-Duplex Omni-Modal Post-Training. This child issue specifies the turn-bounded OPD foundation; streaming full-duplex OPD is tracked in #566. It does not replace the parent's shared compatibility and GRPO validation gates.

**Status:** design proposal. The recipes below are proposed interfaces, not commands already available on `main`. No completed implementation or training results are claimed.

Last updated: 09/08/2026.

### 1. Scope

Here **simplex (单工)** means the official MiniCPM-o `as_simplex()` / half-duplex behavior: consume the complete user turn, then generate the answer. Delivering that answer in chunks does not make the conversation full duplex.

| Stage | Inputs / conditioning | Distilled policy |
| --- | --- | --- |
| Thinker, first milestone | Complete text, then image/audio and bounded audio-video inputs | Assistant text tokens |
| Talker, subsequent milestone | Student-generated text, native LLM-state handoff, and codec history | AR speech-codec tokens |
| Code2wav | Codec tokens | Frozen waveform rendering; no AR policy loss |

Train **one AR stage per job** initially, with stage-specific LoRA and frozen vision/audio encoders. Thinker training may retain a frozen speech pipeline for evaluation. **Audible output is not evidence that the talker policy was distilled.** For talker training, freeze the thinker and retain its native conditioning.

Non-goals are simultaneous multi-stage optimization, cross-tokenizer distillation, learning duplex behavior from a non-duplex checkpoint, and persistent streaming sessions. Cross-tokenizer OPD is tracked by #382.

### 2. Reuse and relationship to existing work

The design baseline is `verl-omni@a7031c69571e080a364faeb9d6c7f443435e49f9`, with the repository pins:

- vLLM-Omni `ded8934626aaad1a3e816c3a1d9d742efc012d93` and vLLM 0.28.0;
- verl `fefb080262e1c015a0ea05f958822a6a512dc795`;
- MiniCPM-o checkpoint/remote code `503e754207c94da6bb26850b4469f367c9ea3582`.

| Existing work | Reuse / boundary |
| --- | --- |
| #345 | Shared MiniCPM compatibility, bounded action contract, and validation roadmap; this issue adds the simplex OPD-specific teacher/loss/recipe integration |
| #406 | Thinker-only text GRPO baseline; not a second implementation of that baseline |
| Merged #527 | `OmniModelBase`, multi-stage rollout hooks, selected policy stage, and stage-specific weight sync |
| Open #375 | Shared omni teacher replica, prompt-logprob export, and teacher-field padding; coordinate rather than introducing another teacher manager |
| Open #550 | MiniCPM understanding-only offline DPO: share model loading and image/audio processing where applicable, but online OPD and speech-policy replay are separate |

The DMD-family controller/runtime work in #545/#546 is not the AR OPD route and is not a dependency of this proposal.

Follow the [omni model integration guide](https://github.com/verl-project/verl-omni/blob/a7031c69571e080a364faeb9d6c7f443435e49f9/docs/contributing/integrating_an_omni_model.md). Keep model-specific loading, processor handling, and native forward preparation behind the adapter; reuse the FSDP engine and teacher resource pool.

The official model card recommends Transformers 4.51.0, whereas this repository targets Transformers `>=5.13.0,<=5.14.1`. Differentiable remote-code forward, processing, FSDP, LoRA, and sync on the repository stack are an explicit compatibility gate. Do not silently downgrade the shared stack or assume serving compatibility proves training compatibility.

### 3. Training and probability contract

```text
complete prompt + media
  → student rollout, selected AR stage
  → native sampled IDs + behavior log-probs + replay payload
  → frozen teacher scores those same IDs at the same prefixes
  → actor recomputes current log-probs
  → existing OPD loss and optimizer
  → selected-stage weight sync
  → next student rollout
```

Teacher/student requirements:

- Validate identical tokenizer mappings, special tokens, and, for talker jobs, codec vocabulary/format. Matching vocabulary size alone is insufficient.
- Teacher weights remain fixed. A meaningful experiment needs a justified teacher advantage, such as a stronger compatible fine-tuned checkpoint or a supported capacity-reduced student. Using the same checkpoint for both is a consistency baseline, not a learning result.
- Preserve the student token IDs; do not decode/re-tokenize its answer or replace its history with a teacher-generated answer.
- Match multimodal expansion, positions, prefix lengths, and action masks across rollout, actor, and teacher. Teacher context budgets include the complete processed prefix plus student response and scoring headroom.
- Frozen teacher state is separate from actor/reference-policy state and never receives student weight updates.

Start with verl's sampled reverse-KL policy-gradient path:

```yaml
distillation:
  enabled: true
  distillation_loss:
    loss_mode: kl
    use_policy_gradient: true
    use_task_rewards: false
    policy_loss_mode: vanilla
```

The target is `KL(student || teacher)` on student-visited prefixes. The existing loss uses `d_i = log p_student(a_i | h_i) - log p_teacher(a_i | h_i)` and supplies `-stop_gradient(d_i)` to the policy-gradient loss with the current/old-policy ratio and valid-action mask. Reuse that implementation; directly differentiating the fixed sampled log-probability difference is not the same update.

Use temperature 1, no optional top-k/top-p truncation or repetition penalty, and one PPO epoch for the alignment baseline. Native codec constraints must be preserved. Declare and reproduce any mandatory probability processing rather than mixing processed rollout probabilities with raw teacher/actor probabilities. Forced tokens and padding have no policy loss.

Top-k forward KL and task rewards are subsequent options after their alignment paths are tested; top-k KL must be described as an approximation, not a full-vocabulary KL.

### 4. Thinker and talker replay

**Thinker:** use the MiniCPM multimodal forward, including modality bounds and positions. Calling only the underlying LLM is valid for a verified text-only input, not an arbitrary image/audio/video example. Teacher scoring should reuse #375's omni integration after reconciliation with #527.

**Talker:** codec likelihood depends on more than text or codec IDs. Retain text-token spans, required LLM states, codec history, and native chunk/end-of-turn boundaries. Expose a differentiable codec-likelihood forward, not an inference-only `generate_chunk()` or waveform loss.

If teacher/student upstream components are identical and frozen, their handoff can be reused after validating provenance. Otherwise, reconstruct the teacher's own handoff from the same input and student actions; feeding student hidden states into an incompatible teacher silently changes the objective. Cached hidden states cannot replace recomputation through an upstream component being updated.

The initial recipe may collect a complete bounded response before scoring. Chunked output is supported only when all policy-stage tokens and handoffs survive collection. Missing replay data must fail explicitly, not fall back to text-only training.

### 5. Weight preparation and data

Prepare the official checkpoint and review its remote model code:

```bash
export BASE_MODEL="$HOME/models/MiniCPM-o-4_5"
hf download openbmb/MiniCPM-o-4_5 \
  --revision 503e754207c94da6bb26850b4469f367c9ea3582 \
  --local-dir "$BASE_MODEL"

export STUDENT_MODEL="$BASE_MODEL"
export TEACHER_MODEL="$HOME/models/minicpm-o45-teacher"
```

`TEACHER_MODEL` must point to an actual compatible frozen checkpoint; no stronger public teacher is assumed. For speech, retain the tokenizer, processor, TTS and `assets/token2wav/` assets. Follow the [repository installation guide](https://github.com/verl-project/verl-omni/blob/a7031c69571e080a364faeb9d6c7f443435e49f9/docs/start/install.md) and the pinned runtime in an isolated environment.

Reuse structured messages and media columns. A proposed source JSONL record, before Parquet conversion, is:

```json
{"data_source":"minicpm_o45_simplex","prompt":[{"role":"user","content":"<image>\n<audio>\nAnswer the spoken question about this scene."}],"images":[{"path":"images/scene.png"}],"audios":[{"path":"audio/question.wav","sample_rate":16000}],"extra_info":{"sample_id":"scene-001"}}
```

The converter preserves ordered media slots, sample rates, and, for video, frame timestamps and soundtrack alignment. Voice-reference audio is explicit speaker conditioning, not an extra user-audio slot. No chosen/rejected pair or teacher answer is required: responses are sampled from the student online.

Start with text prompts, then licensed image/audio QA and bounded AV tasks. The AVQA investigation in #345 can inform the latter, without treating AVQA as conversational/duplex data. Keep evaluation splits and answer annotations out of teacher context. Dataset adoption must check source-media rights and speaker consent.

### 6. Proposed usage

**These scripts and wrapper variables are to be implemented:**

```bash
python examples/opd_trainer/minicpm_o/prepare_data.py \
  --mode simplex --input train.jsonl --output "$HOME/data/minicpm_o45_simplex"

STUDENT_MODEL="$STUDENT_MODEL" TEACHER_MODEL="$TEACHER_MODEL" \
DATA_DIR="$HOME/data/minicpm_o45_simplex" TRAIN_STAGE=thinker \
STUDENT_GPUS=4 TEACHER_GPUS=4 \
bash examples/opd_trainer/minicpm_o/run_simplex_opd_lora.sh
```

`TRAIN_STAGE=talker` is enabled only after codec replay and scoring pass validation. It is not an alias for audible output.

The recipe should use `python -m verl_omni.trainer.main_omni` with `VERL_USE_EXTERNAL_MODULES=verl_omni`, the existing `actor_rollout_ref.model.path` / `model_stage`, FSDP2 and LoRA settings, `rollout.name=vllm_omni`, and `distillation.teacher_models.teacher_model` with omni inference. Teacher capacity uses the existing `distillation.nnodes` / `n_gpus_per_node` and inference TP settings.

Student actor/rollout may be colocated; teacher workers use a separate pool. The 4+4 GPU split is a proposed profiling topology, not a measured minimum. Do not reuse diffusion SDE steps, rewards, or group-size assumptions for this AR workload.

### 7. Delivery and acceptance

1. **Compatibility:** share the adapter foundation with #406/#550 and #527; validate selected-stage differentiable forward and sync against the real checkpoint.
2. **Simplex thinker OPD:** text first, then image/audio and bounded AV. Complete rollout → teacher scoring → masked update → sync → fresh rollout.
3. **Simplex talker OPD:** native handoff and codec likelihood, stage-specific teacher scoring, saved speech and selected-stage-only updates.
4. **Behavior and throughput:** several updates, held-out quality, teacher coverage/latency, resource use, and checkpoint/resume at completed-turn boundaries.

Acceptance includes CPU tests for vocabulary/position/mask/transport alignment; fixed-prefix actor/rollout/teacher comparisons with declared dtype tolerances; near-zero signal for identical policies; nonzero intended-stage gradients for a differing teacher; finite log-probabilities; post-sync verification; and text/multimodal/speech regression checks. Lower KL alone is not a quality claim. Teacher or replay failures invalidate the example rather than silently switching to reward-only RL.

Open questions are the teacher checkpoint choice, remote-code training compatibility, native talker likelihood interface, and teacher throughput. This issue narrows the OPD integration work; it does not duplicate or bypass the shared baseline tracked by #345/#406.

## Motivation

Turn-bounded OPD is the smallest useful distillation entry point for MiniCPM-o 4.5. It gives dense supervision on student-generated trajectories without requiring preference pairs or a reward for every token, while separating model/teacher alignment failures from streaming-session complexity.

The resulting selected-stage adapters, bounded replay, and teacher integration also form reusable prerequisites for the streaming full-duplex OPD child issue under #345.

## Your contribution

I propose to contribute the MiniCPM-specific adapter integration, simplex thinker/talker OPD recipes, data conversion, documentation, and validation, coordinating shared teacher and model work with #375/#406/#550.

This RFC was prepared with AI assistance. It describes proposed work, not completed implementation or training results; implementation changes require human review.

### Sources

- [Official MiniCPM-o 4.5 model card and simplex usage](https://huggingface.co/openbmb/MiniCPM-o-4_5/blob/503e754207c94da6bb26850b4469f367c9ea3582/README.md)
- [Released remote model implementation](https://huggingface.co/openbmb/MiniCPM-o-4_5/blob/503e754207c94da6bb26850b4469f367c9ea3582/modeling_minicpmo.py)
- [Pinned vLLM-Omni stage topology](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/model_executor/models/minicpmo_4_5/pipeline.py) and [text-to-speech handoff](https://github.com/vllm-project/vllm-omni/blob/ded8934626aaad1a3e816c3a1d9d742efc012d93/vllm_omni/model_executor/stage_input_processors/minicpmo_4_5_omni.py)
- [verl teacher scoring](https://github.com/verl-project/verl/blob/fefb080262e1c015a0ea05f958822a6a512dc795/verl/experimental/teacher_loop/teacher_manager.py) and [distillation losses](https://github.com/verl-project/verl/blob/fefb080262e1c015a0ea05f958822a6a512dc795/verl/trainer/distillation/losses.py)
